File: android/apps/AutoTesterAndroid/app/src/main/java/com/mousebirdconsulting/autotester/TestCases/ClusteredMarkersTestCase.java
/*
 *  ClusteredMarkersTestCase.java
 *  WhirlyGlobeLib
 *
 *  Created by jmnavarro
 *  Copyright 2011-2016 mousebird consulting
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */
package com.mousebirdconsulting.autotester.TestCases;

import android.app.Activity;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.BitmapFactory;

import com.mousebird.maply.GlobeController;
import com.mousebird.maply.MapController;
import com.mousebird.maply.BaseController;
import com.mousebird.maply.MarkerInfo;
import com.mousebird.maply.Point2d;
import com.mousebird.maply.RenderController;
import com.mousebird.maply.ScreenMarker;
import com.mousebird.maply.VectorObject;
import com.mousebird.maply.MaplyTexture;
import com.mousebird.maply.BasicClusterGenerator;
import com.mousebird.maply.RenderController;
import com.mousebirdconsulting.autotester.Framework.MaplyTestCase;
import com.mousebirdconsulting.autotester.R;

import java.util.ArrayList;
import java.util.List;

public class ClusteredMarkersTestCase extends MaplyTestCase {

    public ClusteredMarkersTestCase(Activity activity) {
        super(activity);
        setTestName("Clustered Markers");
        setDelay(1000);
        this.implementation = TestExecutionImplementation.Both;
    }

    @Override
    public boolean setUpWithMap(MapController mapVC) throws Exception {
        VectorsTestCase baseView = new VectorsTestCase(getActivity());
        baseView.setUpWithMap(mapVC);
        insertClusteredMarkers(baseView.getVectors(), mapVC);
        mapVC.setPositionGeo(pos.getX(), pos.getY(), 2);
        return true;
    }

    @Override
    public boolean setUpWithGlobe(GlobeController globeVC) throws Exception {
        VectorsTestCase baseView = new VectorsTestCase(getActivity());
        baseView.setUpWithGlobe(globeVC);
        insertClusteredMarkers(baseView.getVectors(), globeVC);
        globeVC.animatePositionGeo(pos.getX(), pos.getY(), 0.9, 1);

        globeVC.setPerfInterval(20);

        return true;
    }

    Point2d pos = Point2d.FromDegrees(-3.6704803, 40.5023056);

    private void insertClusteredMarkers(List<VectorObject> vectors, BaseController inController) {
        Point2d size = new Point2d(32, 32);
        List<ScreenMarker> markers = new ArrayList<>();
        Bitmap icon = BitmapFactory.decodeResource(getActivity().getResources(), R.drawable.sticker);
        MaplyTexture tex = inController.addTexture(icon,new RenderController.TextureSettings(), RenderController.ThreadMode.ThreadCurrent);

//        inController.addClusterGenerator(new BasicClusterGenerator(new int[]{Color.argb(165, 255, 255, 0)}, 1, new Point2d(64, 64), inController, inController.getActivity()));

        int which = 0;
        for (VectorObject v : vectors) {
            // Note: Increase this to test capacity
            for (int ii=0;ii<1;ii++) {
                ScreenMarker marker = new ScreenMarker();
                marker.tex = tex;
                marker.loc = v.centroid();
                marker.size = size;
                marker.selectable = true;

                if (marker.loc != null)
                    markers.add(marker);
            }
            which++;
        }

        MarkerInfo info = new MarkerInfo();
        info.setLayoutImportance(1.f);
//        info.setClusterGroup(1);
        info.setClusterGroup(0);

        inController.addScreenMarkers(markers, info, RenderController.ThreadMode.ThreadAny);
    }
}

File: android/apps/AutoTesterAndroid/app/src/main/java/com/mousebirdconsulting/autotester/TestCases/ClusteredMarkersTestCase.kt
/*  ClusteredMarkersTestCase.kt
 *  WhirlyGlobeLib
 *
 *  Created by jmnavarro
 *  Copyright 2011-2021 mousebird consulting
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.mousebirdconsulting.autotester.TestCases

import android.app.Activity
import android.graphics.*
import android.util.Log
import com.mousebird.maply.*
import com.mousebird.maply.RenderControllerInterface.TextureSettings
import com.mousebirdconsulting.autotester.Framework.MaplyTestCase
import java.util.*

class ClusteredMarkersTestCase(activity: Activity?) :
        MaplyTestCase(activity, "Clustered Markers", TestExecutionImplementation.Both) {

    @Throws(Exception::class)
    override fun setUpWithMap(mapVC: MapController): Boolean {
        baseCase.setUpWithMap(mapVC)
        baseCase.baseCase.setForwardMapDelegate(this)  // Get delegate calls from the underlying case
        insertClusteredMarkers(baseCase.vectors, mapVC)
        mapVC.setPositionGeo(pos.x, pos.y, 2.0)
        mapVC.setAllowRotateGesture(true)
        return true
    }
    
    @Throws(Exception::class)
    override fun setUpWithGlobe(globeVC: GlobeController): Boolean {
        baseCase.setUpWithGlobe(globeVC)
        baseCase.baseCase.setForwardGlobeDelegate(this)    // Get delegate calls from the underlying case
        insertClusteredMarkers(baseCase.vectors, globeVC)
        globeVC.animatePositionGeo(pos.x, pos.y, 0.9, 1.0)
        globeVC.keepNorthUp = false
        globeVC.setPerfInterval(20)
        return true
    }
    
    val pos: Point2d = Point2d.FromDegrees(-3.6704803, 40.5023056)

    private fun loadIcon(name: String): Bitmap =
        activity.assets.open("maki icons/$name.png").use { BitmapFactory.decodeStream(it) }
    private fun loadTex(name: String): MaplyTexture? =
        controller.addTexture(loadIcon(name), TextureSettings(), ThreadMode.ThreadCurrent)

    private fun insertClusteredMarkers(vectors: List<VectorObject>, inController: BaseController) {
        val iconSize = Point2d(64.0, 64.0)
        val clusterMarkers: MutableList<ScreenMarker> = ArrayList(vectors.size)
        val pointMarkers: MutableList<ScreenMarker> = ArrayList(vectors.size)
        val markerIcon = loadTex("marker-stroked-24@2x")
        val pointIcon = loadTex("circle-24@2x")
        for (v in vectors) {
            (v.centroid() ?: v.center())?.let { c ->
                pointMarkers.add(ScreenMarker().apply {
                    loc = c
                    tex = pointIcon
                    size = Point2d(8.0, 8.0)
                    selectable = false
                    layoutImportance = Float.MAX_VALUE
                })
                // Note: Increase this to test capacity
                for (ii in 0..0) {
                    clusterMarkers.add(ScreenMarker().apply {
                        loc = c
                        tex = markerIcon
                        size = iconSize
                        selectable = true
                        // Offset based on hemispheres
                        offset = Point2d(iconSize.x * (c.x / Math.PI / 2),
                                      iconSize.y * (c.y / Math.PI))
                        // Some fixed to map rotation, some fixed to screen
                        rotation = if (Math.random() < 0.5) 1.0e-8 else 0.0
                        layoutImportance = ii.toFloat()
                    })
                }
            }
        }
        val info = MarkerInfo().apply {
            setClusterGroup(0)
        }
        inController.addScreenMarkers(clusterMarkers, info, ThreadMode.ThreadAny)?.let {
            objs.add(it)
        }
        inController.addScreenMarkers(pointMarkers, info, ThreadMode.ThreadAny)?.let {
            objs.add(it)
        }
    }
    
    override fun userDidSelect(globeControl: GlobeController?, selObjs: Array<out SelectedObject>?, loc: Point2d?, screenLoc: Point2d?) {
        super.userDidSelect(globeControl, selObjs, loc, screenLoc)
        updateClusterGen()
    }
    override fun userDidTap(globeControl: GlobeController, loc: Point2d, screenLoc: Point2d) {
        super.userDidTap(globeControl, loc, screenLoc)
        updateClusterGen()
    }
    
    override fun userDidSelect(mapControl: MapController?, selObjs: Array<out SelectedObject>?, loc: Point2d?, screenLoc: Point2d?) {
        super.userDidSelect(mapControl, selObjs, loc, screenLoc)
        updateClusterGen()
    }
    override fun userDidTap(mapControl: MapController, loc: Point2d, screenLoc: Point2d) {
        super.userDidTap(mapControl, loc, screenLoc)
        updateClusterGen()
    }
    
    override fun shutdown() {
        controller.removeObjects(objs, ThreadMode.ThreadCurrent)
        objs.clear()
        baseCase.shutdown()
        super.shutdown()
    }
    
    private fun updateClusterGen() {
        controller.clearClusterGenerators()
        n = (n + 1) % 4
        when (n) {
            0 -> {
                Log.d("Maply", "Default Clustering")
                controller.addDefaultClusterGenerator()
            }
            1 -> {
                Log.d("Maply", "Custom 1")
                controller.addClusterGenerator(BasicClusterGenerator(intArrayOf(
                        Color.argb(165, 255, 255, 0)),
                        0, Point2d(64.0, 64.0), controller))
            }
            2 -> {
                Log.d("Maply", "Custom 2")
                controller.addClusterGenerator(BasicClusterGenerator(intArrayOf(
                        Color.argb(255, 0, 0, 0),
                        Color.argb(255, 0, 255, 255),
                        Color.argb(255, 255, 255, 255)),
                        0, Point2d(96.0, 64.0), controller))
            }
            3 -> Log.d("Maply", "No clustering")
        }
    }
    
    private var n = -1
    private val objs = ArrayList<ComponentObject>()
    private var baseCase = VectorsTestCase(activity)
}